### PR TITLE
implement div for keyexpr joining, factorize generics to help with co…

### PR DESCRIPTION
**Pull request by p-avital**
Remove autocanonization from the TryFrom constructor, `autocanonize` constructor replaces it.

Factorize generics.

impl Div for infallible keyexpr joining.

### CODE REVIEW

File: commons/zenoh-protocol-core/src/key_expr/borrowed.rs [hunk 1]
```diff
@@ -249,11 +265,12 @@ impl keyexpr {
...
-    where
-        &'a Self: TryFrom<T, Error = E>,
-    {
-        t.try_into()
+}
+
+impl std::ops::Div for &keyexpr {
+    type Output = OwnedKeyExpr;
+    fn div(self, rhs: Self) -> Self::Output {
+        self.join(rhs).unwrap() // Joining 2 key expressions should always result in a canonizable string.
```
Comment (JEnoch) [inline commons/zenoh-protocol-core/src/key_expr/borrowed.rs:273]: Didn't you forget the call to `Self::autocanonize(s).unwrap()` here (as done in Div for OwnedKeyExpr) ?

File: commons/zenoh-protocol-core/src/key_expr/borrowed.rs [hunk 2]
```diff
@@ -42,6 +42,22 @@ use super::{OwnedKeyExpr, FORBIDDEN_CHARS};
 pub struct keyexpr(str);
 
 impl keyexpr {
+    pub fn new<'a, T, E>(t: &'a T) -> Result<&'a Self, E>
```
Comment (JEnoch) [inline commons/zenoh-protocol-core/src/key_expr/borrowed.rs:45]: Documentation would be welcome 😊 !

File: commons/zenoh-protocol-core/src/key_expr/borrowed.rs [hunk 3]
```diff
@@ -42,6 +42,22 @@ use super::{OwnedKeyExpr, FORBIDDEN_CHARS};
...
+    where
+        &'a Self: TryFrom<&'a T, Error = E>,
+        T: ?Sized,
+    {
+        t.try_into()
+    }
+
+    pub fn autocanonize<'a, T, E>(t: &'a mut T) -> Result<&'a Self, E>
```
Comment (JEnoch) [inline commons/zenoh-protocol-core/src/key_expr/borrowed.rs:53]: Documentation would be welcome 😊 !

File: commons/zenoh-protocol-core/src/key_expr/owned.rs [hunk 4]
```diff
@@ -26,6 +26,21 @@ use std::{convert::TryFrom, str::FromStr};
 pub struct OwnedKeyExpr(pub(crate) Box<str>);
 
 impl OwnedKeyExpr {
+    pub fn new<T, E>(t: T) -> Result<Self, E>
```
Comment (JEnoch) [inline commons/zenoh-protocol-core/src/key_expr/owned.rs:29]: Documentation would be welcome 😊 !

File: commons/zenoh-protocol-core/src/key_expr/owned.rs [hunk 5]
```diff
@@ -26,6 +26,21 @@ use std::{convert::TryFrom, str::FromStr};
...
+    where
+        Self: TryFrom<T, Error = E>,
+    {
+        Self::try_from(t)
+    }
+
+    pub fn autocanonize<T, E>(mut t: T) -> Result<Self, E>
```
Comment (JEnoch) [inline commons/zenoh-protocol-core/src/key_expr/owned.rs:36]: Documentation would be welcome 😊 !

File: zenoh/src/key_expr.rs [hunk 6]
```diff
@@ -193,6 +194,22 @@ impl KeyExpr<'static> {
     }
 }
 impl<'a> KeyExpr<'a> {
+    pub fn new<T, E>(t: T) -> Result<Self, E>
```
Comment (JEnoch) [inline zenoh/src/key_expr.rs:197]: Documentation would be welcome 😊 !

File: zenoh/src/key_expr.rs [hunk 7]
```diff
@@ -193,6 +194,22 @@ impl KeyExpr<'static> {
...
+    where
+        Self: TryFrom<T, Error = E>,
+    {
+        Self::try_from(t)
+    }
+
+    pub fn autocanonize<T, E>(mut t: T) -> Result<Self, E>
```
Comment (JEnoch) [inline zenoh/src/key_expr.rs:204]: Documentation would be welcome 😊 !

**Review by JEnoch — CHANGES_REQUESTED** [review]

File: commons/zenoh-protocol-core/src/key_expr/borrowed.rs (hunk 1 above)
Comment (p-avital) [reply commons/zenoh-protocol-core/src/key_expr/borrowed.rs:273]: I didn't, `join` already does autocanonization, it just doesn't assume that things will work out as we can with a pair or KEs

**Review by JEnoch — APPROVED** [review]
LGTM

**Pull request merged**